libkernelflinger/tpm2_security.c

```diff
@@ -808,6 +828,118 @@ EFI_STATUS tpm2_fuse_bootloader_policy(
 	return EFI_UNSUPPORTED;
 }
 
+static EFI_STATUS tpm2_fuse_bootloader(void)
+{
+	EFI_STATUS ret;
+	UINT16 config_index;
+	BYTE data[NV_INDEX_BOOTLOADER_SIZE];
+	BYTE data_read[NV_INDEX_BOOTLOADER_SIZE];
+	UINT16 data_read_size = NV_INDEX_BOOTLOADER_SIZE;
+	tpm2_bootloader_t *bootloader = (tpm2_bootloader_t *)data;
+
+	config_index = NV_INDEX_BOOTLOADER - config_table[0].nv_index;
+	ret = tpm2_create_nvindex(NV_INDEX_BOOTLOADER, config_table[config_index].attribute, NV_INDEX_BOOTLOADER_SIZE);
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Failed to create bootloader NV index");
+		return ret;
+	}
+
+	memset(data, 0, sizeof(data));
+	bootloader->struct_ver = NV_INDEX_BOOTLOADER_STRUCT_VER;
+	bootloader->lock_state = JUST_INIT;
+
+	// Read the data again to verify it
+	ret = tpm2_write_nvindex(NV_INDEX_BOOTLOADER, NV_INDEX_BOOTLOADER_SIZE, data, 0);
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Write bootloader NV index failed");
+		goto out;
+	}
+
+	// Read the data again to verify it
+	ret = tpm2_read_nvindex(NV_INDEX_BOOTLOADER, &data_read_size, data_read, 0);
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Read bootloader NV index back failed just after write it");
+		goto out;
+	}
+
+	if (memcmp(data, data_read, sizeof(data))) {
+		error(L"Security error! Read bootloader NV index back but verify failed!");
+		ret = EFI_SECURITY_VIOLATION;
+		goto out;
+	}
+
+	debug(L"Success create and write bootloader NV index");
+out:
+	return ret;
+}
+
+static EFI_STATUS tpm2_check_bootloader_index(void)
+{
+	EFI_STATUS ret;
+	TPM2B_NV_PUBLIC NvPublic;
+	TPM2B_NAME NvName;
+	UINT8 struct_ver;
+	UINT16 data_size = sizeof(struct_ver);
+
+	// Check the SEED nvindex
+	ret = Tpm2NvReadPublic(NV_INDEX_BOOTLOADER, &NvPublic, &NvName);
+	if (!EFI_ERROR(ret)) {
```

Comment:
You should try to reverse the condition to have most of the code at the root of the function and reduce the indentation.